### PR TITLE
feat(authmethods): Type Authenticate request and response attributes

### PR DESCRIPTION
**Pull request by johanbrandhorst**
## refact(subtype): Remove copied structToProto methods

The original idea was for these methods to eventually become
entirely encapsulated by the subtype package, but as it turns
out we still have use for them in custom transformation functions,
so move the subtype package to use these shared functions
instead.

## feat(authtoken): Add token type to AuthToken

The token type is used in the response from the OIDC
Authenticate method only.

## feat(authmethods): Add typed attributes to authenticate request and response

The AuthenticateRequest and AuthenticateResponse types have a
complex mapping rule system, which requires a pair of custom
callback functions.

### CODE REVIEW

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 1]
```diff
@@ -1194,11 +1199,103 @@ func (s Service) convertToAuthenticateResponse(ctx context.Context, req *pbs.Aut
...
+
+	return &pbs.AuthenticateResponse{
+		Command: req.GetCommand(),
+		Attrs: &pbs.AuthenticateResponse_AuthTokenAttributes{
+			AuthTokenAttributes: tok,
+		},
+	}, nil
+}
+
+func transformAuthenticateRequestAttributes(msg proto.Message) error {
```
Comment (tmessi) [inline internal/servers/controller/handlers/authmethods/authmethod_service.go:1212]: can you add unit tests for these two transform functions?

**Review by tmessi — COMMENTED** [review]
looks good. just wondering if there can be some new tests added

**Review by tmessi — APPROVED** [review]

File: internal/proto/controller/api/resources/authtokens/v1/authtoken.proto [hunk 2]
```diff
@@ -42,6 +42,9 @@ message AuthToken {
 	// Output only. The time this Auth Token expires.
 	google.protobuf.Timestamp expiration_time = 110 [json_name="expiration_time"];
 
+	// Output only. The type of this auth token. Either "cookie" or "token".
+	string token_type = 120 [json_name="token_type"];
```
[the remark this answers is not in the document]
Comment (tmessi) [reply internal/proto/controller/api/resources/authtokens/v1/authtoken.proto:46]: Yea, if we can avoid any external API changes I think we should, even though this is additive.

[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply internal/proto/controller/api/resources/authtokens/v1/authtoken.proto]: Duplicated the type in 76a9543c5a0454cf5cf69a163ead89ed949d0114

**Review by tmessi — APPROVED** [review]

**Pull request merged**